Fix: urls for user details

parasharrajat opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This issue was caused due to the reason that Url Path had `.` which is not being parsed in React-navigation. I have also reported the issue to React-navigation but nobody is even bothering to reply so we have to fix it with a patch in our app. I will submit a new PR as soon as I got some leads on that issue which is very unlikely.

To fix, PR does the following:
1. Covert the route param `:login` to query param `?:login=xxx` which fixes the issue. and routes are parsed successfully.
e.g. `https://staging.expensify.cash/details/:login` to `https://staging.expensify.cash/details?login=:login`


### Fixed Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing -->
$ Fixes #3924

### Tests | QA Steps
1. Open Any chat in E.cash on web.
2. Click user avatar to see its details.
3. refresh the page, you should land on the same page and app should not crash.

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

https://user-images.githubusercontent.com/24370807/127411771-a160378f-4966-44ae-8566-75d60f440c0b.mp4


#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

parasharrajat committed:
fix: urls for user details

chiragsalian commented on src/ROUTES.js:
```diff
@@ -96,4 +96,22 @@ export default {
+
+    /**
+     * React Navigation is failing to parse urls with dot in the path segment.
+     * This method enable pasring of urls with dot in path Segment
```
NAB: switch 'enable' with 'enables'.
switch 'pasring' with 'parsing'.

chiragsalian commented on src/ROUTES.js:
```diff
@@ -96,4 +96,22 @@ export default {
+     * @param {String} path - Path for the config
+     * @returns {Object}
+     */
+    getEmailRouteLinkingConfig: path => ({
```
In the PR details you explain why you moved away from the initial solution but i don't quite understand the current solution. Can you add it to your PR or explain it to me.

parasharrajat replied:
@chiragsalian  We can customize the default param parsing by using the parse and stringify option for LinkingConfig.
1. In stringify, I replace all the dots with `@dot@` safely. we first encode the existing `@dot@` in the URL and then do the dot conversion. Which will allow `@dot@` to be part of the email.
2. In parse, we revert to the original email by first converting `@dot@` back to dot and then replace the escaped version of `@dot@` back to `@dot@`.

parasharrajat replied:
Any thoughts @chiragsalian.

chiragsalian replied:
Ah nice, i had no idea. I just read more about this [here](https://reactnavigation.org/docs/configuring-links/#passing-params) too. Well it makes sense to me and i think its fine, but i'll loop in @marcaaron too for additional thoughts.

parasharrajat replied:
@marcaaron Do you have any thoughts on this issue?

parasharrajat replied:
Hey @marcaaron What do you think about it? Sorry for the ping.

marcaaron commented on src/ROUTES.js:
```diff
@@ -96,4 +96,22 @@ export default {
+        path,
+        parse: {
+            login: l => (l
+                ? decodeURIComponent(l).replace('@dot@', '.').replace(encodeURIComponent('@dot@'), '@dot@')
```
What does the second `.replace()` achieve here? Didn't we replace with `.` already?

parasharrajat replied:
I added this for safe parsing. In case the email already contains `@dot@` e.g. `rajat"@dot@"parashar@gmail.com`. Please let me know if I have forgotten to encounter any special cases.

`@dot@` I just took it based on the meaning. dot for `.` and `@` is a special URL character that allows us to safely convert emails that can contain `@dot@` as part of the string. 
`@` will be escaped and thus we can use it.

marcaaron replied:
>In case the email already contains @dot@ e.g. rajat"@dot@"parashar@gmail.com. Please let me know if I have forgotten to encounter any special cases.

Sorry, not following. Why would a valid email have more than one `@` let alone an `@dot@` in it?

parasharrajat replied:
Ok. I saw somewhere that @ can be used in quotes. I believe I overthink it. I 'll remove the extra escaping.

parasharrajat replied:
@marcaaron Updated the PR. Removed the Extra escaping stuff.

marcaaron commented on src/ROUTES.js:
```diff
@@ -96,4 +96,22 @@ export default {
+                : undefined),
+        },
+        stringify: {
+            login: l => (l ? l.replace('@dot@', encodeURIComponent('@dot@')).replace('.', '@dot@') : undefined),
```
The logic here is also confusing... 

```js
l
    .replace('@dot@', encodeURIComponent('@dot@'))
    .replace('.', '@dot@')
```

- We are replacing the `@dot@` with `%40dot%40`
- Then we replace all `.` with `@dot@`... 

Why?

parasharrajat replied:
Maybe, I have answered it above. First replace will escape the existing `@dot@` from email and then we can use `@dot@` to convert `.`.

marcaaron commented:
@parasharrajat Is there some more context on the `@dot@` addition? URLs are going to look pretty weird after we do this no? Just wondering if we checked with anyone whether the weird looking urls are what we want.

parasharrajat commented:
I checked out on google, Using emails as URL is not a recommended thing as emails are considered private data. So if we use them in URL then it is kind of public. But we are not using any userIds so It's kind of good that the URL does not look like containing emails and bots can't read emails from it.
Thoughts?

marcaaron commented:
>I checked out on google, Using emails as URL is not a recommended thing as emails are considered private data. So if we use them in URL then it is kind of public. But we are not using any userIds so It's kind of good that the URL does not look like containing emails and bots can't read emails from it.

- The emails are easily accessible. What do we achieve by obscuring them in the URL only?
- The app is used to converse with people you know and work with not random people. Maybe that will change in the future, but it's not a problem today.

>we are not using any userIds so It's kind of good that the URL does not look like containing emails and bots can't read emails from it

I'm not sure this is a problem we need to solve right now, but could also not be understanding which bots we should be worried about.

To clarify, my concern is that the url looks weird and I don't see any discussion about how we landed on `@dot@` or other alternatives. Is there another way to do this? e.g. query params vs. path?

parasharrajat committed:
remove unnecessary escaping

parasharrajat commented:
Yeah, we can do this we Query param. This was the original Idea. But Route will not be strictly matched with the Query param.
for example, /details is equal to /details?user=xxxx;

We can definitely use query params. Let me know? Thanks.

marcaaron commented:
We could maybe bring this up in Slack? I feel others may have opinions to share.

parasharrajat committed:
use query params for login routes

parasharrajat commented:
Ok @marc after discussion https://expensify.slack.com/archives/C01GTK53T8Q/p1627942434082700. I have used query params.

parasharrajat commented:
@marcaaron This is ready. Please let me know if there is something that needs to be changed.

marcaaron commented:
Thanks for the bump! Can we fix up the description so it reflects the final decision that was made?

parasharrajat commented:
Done. Thanks for mentioning that.

chiragsalian commented:
Did you retest? The moment i click on an avatar i get an error,
![image](https://user-images.githubusercontent.com/14356145/128108861-b39ac594-d364-44ba-8d42-af0f90e212f6.png)

Okay looks like normal emails work. I got the failure with the email chirag+15@expensify.com which is a pretty normal test email to have. Would you be able to address this?

parasharrajat committed:
esacpe chars in param

parasharrajat commented:
@chiragsalian Yes I retested it but I didn't test the URL with Plus. I have fixed this issue. 
I think escaping is not happening for query params so I have added it myself and tested it.

chiragsalian reviewed (approved):
Nice, LGTM and works great 👍 
@marcaaron all yours.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1103643857) to staging         in version: 1.0.82-8🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1120793764) to production by @francoisl in version: 1.0.83-1 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|failure ❌
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
